hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MethodProbe.java

```diff
@@ -93,61 +173,68 @@ static <S> MethodProbe createMethodProbe(Method method, Probe probe, SourceMetad
     }
 
     static class LongMethodProbe<S> extends MethodProbe implements LongProbeFunction<S> {
-
-        LongMethodProbe(Method method, Probe probe, int type, SourceMetadata sourceMetadata) {
+        LongMethodProbe(final Method method, final Probe probe, final ProbeType type, final SourceMetadata sourceMetadata) {
             super(method, probe, type, sourceMetadata);
         }
 
+        @SuppressWarnings("checkstyle:CyclomaticComplexity")
         @Override
-        public long get(S source) throws Exception {
+        public long get(final S source) throws Throwable {
             switch (type) {
-                case TYPE_PRIMITIVE_LONG:
-                    return ((Number) method.invoke(source, EMPTY_ARGS)).longValue();
+                case TYPE_LONG_PRIMITIVE:
+                    return isMethodStatic ? (long) methodHandle.invokeExact() : (long) methodHandle.invokeExact(source);
                 case TYPE_LONG_NUMBER:
-                    Number longNumber = (Number) method.invoke(source, EMPTY_ARGS);
+                    final Number longNumber = invoke(source);
                     return longNumber == null ? 0 : longNumber.longValue();
                 case TYPE_MAP:
-                    Map<?, ?> map = (Map<?, ?>) method.invoke(source, EMPTY_ARGS);
+                    final Map<?, ?> map = invoke(source);
                     return map == null ? 0 : map.size();
                 case TYPE_COLLECTION:
-                    Collection<?> collection = (Collection<?>) method.invoke(source, EMPTY_ARGS);
+                    final Collection<?> collection = invoke(source);
                     return collection == null ? 0 : collection.size();
                 case TYPE_COUNTER:
-                    Counter counter = (Counter) method.invoke(source, EMPTY_ARGS);
+                    final Counter counter = invoke(source);
                     return counter == null ? 0 : counter.get();
                 case TYPE_SEMAPHORE:
-                    Semaphore semaphore = (Semaphore) method.invoke(source, EMPTY_ARGS);
+                    final Semaphore semaphore = invoke(source);
                     return semaphore == null ? 0 : semaphore.availablePermits();
                 default:
-                    throw new IllegalStateException("Unrecognized type:" + type);
+                    throw new IllegalStateException("Unrecognized type: " + type);
             }
         }
     }
 
     static class DoubleMethodProbe<S> extends MethodProbe implements DoubleProbeFunction<S> {
-
-        DoubleMethodProbe(Method method, Probe probe, int type, SourceMetadata sourceMetadata) {
+        DoubleMethodProbe(final Method method, final Probe probe, final ProbeType type, final SourceMetadata sourceMetadata) {
             super(method, probe, type, sourceMetadata);
         }
 
         @Override
-        public double get(S source) throws Exception {
+        public double get(final S source) throws Throwable {
             switch (type) {
```

Comment:
Here is the same problem as with the long. The float type is missing. So there needs to be support for a primitive float so that boxing is prevented.

Reply:
> Here is the same problem as with the long. The float type is missing. So there needs to be support for a primitive float so that boxing is prevented.

I don't think this is the case? A `float` will be cast to a `double`, not boxed, and that doesn't have any (significant) performance impact.

Using the same benchmarks as before, a `LongProbe` accessing methods returning `long`, `int`, `Long` & `Integer`:
```
Benchmark                                                      (field)  Mode   Cnt   Score    Error   Units

MethodProbeBenchmark.benchmark                              longMethod  avgt  1000   5.636 ±  0.008   ns/op
MethodProbeBenchmark.benchmark:gc.alloc.rate                longMethod  avgt  1000   0.046 ±  0.006  MB/sec
MethodProbeBenchmark.benchmark:gc.alloc.rate.norm           longMethod  avgt  1000  ≈ 10⁻⁴             B/op

MethodProbeBenchmark.benchmark                               intMethod  avgt  1000   5.643 ±  0.006   ns/op
MethodProbeBenchmark.benchmark:gc.alloc.rate                 intMethod  avgt  1000   0.046 ±  0.006  MB/sec
MethodProbeBenchmark.benchmark:gc.alloc.rate.norm            intMethod  avgt  1000  ≈ 10⁻⁴             B/op

MethodProbeBenchmark.benchmark                        longMethodObject  avgt  1000   4.082 ±  0.055   ns/op
MethodProbeBenchmark.benchmark:gc.alloc.rate          longMethodObject  avgt  1000   0.046 ±  0.006  MB/sec
MethodProbeBenchmark.benchmark:gc.alloc.rate.norm     longMethodObject  avgt  1000  ≈ 10⁻⁴             B/op

MethodProbeBenchmark.benchmark                     integerMethodObject  avgt  1000   3.833 ±  0.010   ns/op
MethodProbeBenchmark.benchmark:gc.alloc.rate       integerMethodObject  avgt  1000   0.046 ±  0.006  MB/sec
MethodProbeBenchmark.benchmark:gc.alloc.rate.norm  integerMethodObject  avgt  1000  ≈ 10⁻⁴             B/op
```

There's no (significant) difference in memory allocation or execution time suggesting that not having an `IntProbe` for example is causing an issue.

Reply:
Normally a float gets cast to a double, but we are talking about signature polymorphism here; so it isn't normal Java code.

This is a JMH benchmark that shows for both an int/float type if there is an upcast to either long/double.

```
package examples;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Measurement;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.annotations.Warmup;

import java.lang.invoke.MethodHandles;
import java.lang.invoke.VarHandle;
import java.util.concurrent.TimeUnit;

@OutputTimeUnit(TimeUnit.NANOSECONDS)
@BenchmarkMode(Mode.AverageTime)
@State(Scope.Benchmark)
@Fork(value = 2)
@Measurement(iterations = 2)
@Warmup(iterations = 1)
public class VarHandleMismatchBenchmark {

    private static final VarHandle INT_VALUE;
    private static final VarHandle FLOAT_VALUE ;

    static {
        try {
            MethodHandles.Lookup l = MethodHandles.lookup();
            INT_VALUE = l.findVarHandle(VarHandleMismatchBenchmark.class, "intValue", int.class);
            FLOAT_VALUE = l.findVarHandle(VarHandleMismatchBenchmark.class, "floatValue", float.class);
        } catch (ReflectiveOperationException e) {
            throw new ExceptionInInitializerError(e);
        }
    }

    public int intValue;
    public float floatValue;

    @Benchmark
    public int int_ExactMatch() {
        return (int) INT_VALUE.get(this);
    }

    @Benchmark
    public long int_upcastToLong() {
        return (long) INT_VALUE.get(this);
    }

    @Benchmark
    public float float_ExactMatch() {
        return (float) FLOAT_VALUE.get(this);
    }

    @Benchmark
    public double float_upcastToDouble() {
        return (double) FLOAT_VALUE.get(this);
    }
}
```

From the performance results it is clear that an upcast isn't treated the same as an exact match.
```Benchmark                                        Mode  Cnt  Score   Error  Units
VarHandleMismatchBenchmark.float_ExactMatch      avgt    4  0.396 ± 0.349  ns/op
VarHandleMismatchBenchmark.float_upcastToDouble  avgt    4  8.829 ± 1.180  ns/op
VarHandleMismatchBenchmark.int_ExactMatch        avgt    4  0.396 ± 0.332  ns/op
VarHandleMismatchBenchmark.int_upcastToLong      avgt    4  8.941 ± 1.271  ns/op
```

Reply:
> This is a JMH benchmark that shows for both an int/float type if there is an upcast to either long/double.

See my [previous comment](https://github.com/hazelcast/hazelcast/pull/25279#discussion_r1336864447), I don't think this issue exists with my implementation using `MethodHandle`s. But this does need to be addressed [when we try to use `FieldProbe`s with `VarHandles`](https://github.com/hazelcast/hazelcast/issues/25528).